File: _shared/typescript/LinkedList.ts
export class ListNode {
  val: number
  next: ListNode | null

  constructor(val?: number, next?: ListNode | null) {
    this.val = (val === undefined ? 0 : val)
    this.next = (next === undefined ? null : next)
  }

  toArray(): Array<number> {
    let head: ListNode | null = this;
    let list: Array<number> = [];
    while (head !== null) {
      list.push(head.val);
      head = head.next;
    }

    return list;
  }
}

export function arrayToListNode(nums: number[]): ListNode | null {
  let current: ListNode | null = null;
  let next: ListNode | null = null;
  for (let i = nums.length - 1; i >= 0; i--) {
    current = new ListNode(nums[i]);
    current.next = next;
    next = current;
  }

  return current;
}

File: _shared/python/linked_list.py
from typing import List, Optional


class ListNode:
    def __init__(self, val: int = 0, next_node: Optional['ListNode'] = None):
        self.val: int = val
        self.next: Optional['ListNode'] = next_node

    def to_list(self) -> List[int]:
        head = self
        output = []
        while head is not None:
            output.append(head.val)
            head = head.next

        return output


def list_to_list_node(nums: List[int]) -> Optional[ListNode]:
    current = None
    next_node = None
    for i in range(len(nums) - 1, -1, -1):
        current = ListNode(nums[i])
        current.next = next_node
        next_node = current

    return current
